File: freightmas/freightmas/report/quotation_list/quotation_list.js
// Copyright (c) 2026, Zvomaita Technologies (Pvt) Ltd and contributors
// For license information, please see license.txt

frappe.query_reports["Quotation List"] = {
	"filters": [
		{
			fieldname: "date_range",
			label: __("Date Range"),
			fieldtype: "Select",
			options: [
				"",
				"Today",
				"Yesterday",
				"This Week",
				"Last Week",
				"This Month",
				"Last Month",
				"This Year",
				"Last Year",
				"Custom"
			],
			default: "This Month",
			on_change: function() {
				let date_range = frappe.query_report.get_filter_value('date_range');
				let today = frappe.datetime.get_today();
				let from_date, to_date;

				switch(date_range) {
					case "Today":
						from_date = to_date = today;
						break;
					case "Yesterday":
						from_date = to_date = frappe.datetime.add_days(today, -1);
						break;
					case "This Week":
						from_date = frappe.datetime.week_start(today);
						to_date = frappe.datetime.week_end(today);
						break;
					case "Last Week":
						let last_week_start = frappe.datetime.add_days(frappe.datetime.week_start(today), -7);
						from_date = last_week_start;
						to_date = frappe.datetime.add_days(last_week_start, 6);
						break;
					case "This Month":
						from_date = frappe.datetime.month_start(today);
						to_date = frappe.datetime.month_end(today);
						break;
					case "Last Month":
						let last_month = frappe.datetime.add_months(today, -1);
						from_date = frappe.datetime.month_start(last_month);
						to_date = frappe.datetime.month_end(last_month);
						break;
					case "This Year":
						let current_year = frappe.datetime.year_start(today);
						from_date = current_year;
						to_date = frappe.datetime.year_end(today);
						break;
					case "Last Year":
						let last_year = new Date(today.split('-')[0] - 1, 0, 1).getFullYear();
						from_date = `${last_year}-01-01`;
						to_date = `${last_year}-12-31`;
						break;
					default:
						from_date = frappe.query_report.get_filter_value('from_date');
						to_date = frappe.query_report.get_filter_value('to_date');
				}

				if (date_range && date_range !== "Custom" && date_range !== "") {
					frappe.query_report.set_filter_value('from_date', from_date);
					frappe.query_report.set_filter_value('to_date', to_date);
				}
			}
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.month_start(),
			reqd: 1
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1
		},
		{
			fieldname: "customer",
			label: __("Customer"),
			fieldtype: "Link",
			options: "Customer",
			only_select: true
		},
		{
			fieldname: "status",
			label: __("Status"),
			fieldtype: "Select",
			options: ["", "Draft", "Pending Approval", "Approved", "Sent to Customer", "Accepted", "Rejected", "Expired", "Cancelled"]
		},
		{
			fieldname: "customer_reference",
			label: __("Reference"),
			fieldtype: "Data"
		}
	],

	onload: function(report) {
		// Add New Quotation button
		report.page.set_primary_action(__('Add Quotation'), function() {
			frappe.new_doc('Quotation');
		});

		// Add Export buttons
		report.page.add_inner_button('Export to Excel', function() {
			const filters = report.get_filter_values(true);
			const query = encodeURIComponent(JSON.stringify(filters));
			const url = `/api/method/freightmas.api.export_report_to_excel?report_name=Quotation List&filters=${query}`;
			window.open(url);
		}, 'Export');

		report.page.add_inner_button('Export to PDF', function() {
			const filters = report.get_filter_values(true);
			const query = encodeURIComponent(JSON.stringify(filters));
			const url = `/api/method/freightmas.api.export_report_to_pdf?report_name=Quotation List&filters=${query}`;
			window.open(url);
		}, 'Export');

		report.page.add_inner_button('Clear Filters', function() {
			// Clear each filter to its default value
			report.filters.forEach(filter => {
				let default_value = filter.df.default || "";
				report.set_filter_value(filter.df.fieldname, default_value);
			});
			
			// Trigger report refresh after clearing filters
			report.refresh();
		});
	}
};

File: freightmas/freightmas/report/quotation_list/quotation_list.py
# Copyright (c) 2026, Zvomaita Technologies (Pvt) Ltd and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import formatdate


def execute(filters=None):
    
    if not filters:
        filters = {}

    columns = get_columns()
    data = []

    # Build conditions and parameters for parameterized query
    conditions = []
    params = {}
    
    if filters.get("from_date"):
        conditions.append("transaction_date >= %(from_date)s")
        params["from_date"] = filters["from_date"]
    
    if filters.get("to_date"):
        conditions.append("transaction_date <= %(to_date)s")
        params["to_date"] = filters["to_date"]
    
    if filters.get("customer"):
        conditions.append("party_name = %(customer)s")
        params["customer"] = filters["customer"]
    
    if filters.get("status"):
        conditions.append("status = %(status)s")
        params["status"] = filters["status"]
    
    if filters.get("job_type"):
        conditions.append("job_type = %(job_type)s")
        params["job_type"] = filters["job_type"]

    where_clause = " AND ".join(conditions) if conditions else "1=1"

    # Handle pagination
    limit_clause = ""
    if filters.get("page_length") and filters.get("page_length") != "All":
        page_length = int(filters.get("page_length", 20))
        start = int(filters.get("start", 0))
        limit_clause = f" LIMIT {page_length} OFFSET {start}"

    # Get quotation data
    quotations = frappe.db.sql("""
        SELECT name, transaction_date, party_name as customer, customer_reference,
               valid_till, job_type, grand_total, status
        FROM `tabQuotation`
        WHERE {where_clause}
        ORDER BY transaction_date DESC
        {limit_clause}
    """.format(where_clause=where_clause, limit_clause=limit_clause), params, as_dict=True)

    for quote in quotations:
        data.append({
            "id": quote.name,
            "transaction_date": format_date(quote.get("transaction_date")),
            "customer": quote.get("customer", ""),
            "customer_reference": quote.get("customer_reference", ""),
            "valid_till": format_date(quote.get("valid_till")),
            "job_type": quote.get("job_type", ""),
            "grand_total": quote.get("grand_total", 0),
            "status": quote.get("status", ""),
        })

    # Return data for pagination
    result = {"data": data, "columns": columns}
    
    # Add total count for pagination if limit is applied
    if limit_clause:
        total_count = frappe.db.sql("""
            SELECT COUNT(*) as count
            FROM `tabQuotation`
            WHERE {where_clause}
        """.format(where_clause=where_clause), params, as_dict=True)[0].count
        result["total"] = total_count
    
    return columns, data


def get_columns():
    """Get column definitions for the quotation list."""
    return [
        {"label": "Quotation No", "fieldname": "id", "fieldtype": "Link", "options": "Quotation", "width": 190},
        {"label": "Created", "fieldname": "transaction_date", "fieldtype": "Data", "width": 110},
        {"label": "Customer", "fieldname": "customer", "fieldtype": "Link", "options": "Customer", "width": 250},
        {"label": "Customer Reference", "fieldname": "customer_reference", "fieldtype": "Data", "width": 170},
        {"label": "Valid Till", "fieldname": "valid_till", "fieldtype": "Data", "width": 110},
        {"label": "Job Type", "fieldname": "job_type", "fieldtype": "Data", "width": 120},
        {"label": "Total Amount", "fieldname": "grand_total", "fieldtype": "Currency", "width": 130},
        {"label": "Status", "fieldname": "status", "fieldtype": "Data", "width": 130},
    ]


def format_date(date_str):
    """Format date consistently."""
    if not date_str:
        return ""
    return formatdate(date_str, "dd-MMM-yy")
